system/Traits/ConditionalTrait.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Traits;
+
+trait ConditionalTrait
+{
+    /**
+     * Only runs the query when $condition evaluates to true
+     *
+     * @param array|bool|float|int|object|resource|string|null $condition
+     *
+     * @return $this
+     */
+    public function when($condition, callable $callback, ?callable $defaultCallback = null): self
+    {
+        if ($condition) {
+            $callback($this, $condition);
+        } elseif ($defaultCallback) {
+            $defaultCallback($this);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Only runs the query when $condition evaluates to false
+     *
+     * @param array|bool|float|int|object|resource|string|null $condition
+     *
+     * @return $this
+     */
+    public function whenNot($condition, callable $callback, ?callable $defaultCallback = null): self
+    {
+        if (! $condition) {
+            $callback($this, $condition);
+        } elseif ($defaultCallback) {
+            $defaultCallback($this);
+        }
+
+        return $this;
```

Comment:
```suggestion
        return $this->whenNot(! $condition, $callback, $defaultCallback);
```

Reply:
`$this->when(! ...)`

Reply:
Unfortunately it's not that simple. All of the tests passed except the last one, `testWhenNotPassesParemeters`. This changes the value of `$condition`, which is passed as a parameter into the value. In this test's case it changed the `0` into `true` when passed into the `when` command. 

While this example is a trivial example of it the issue, it could be a much bigger issue with different data passed into it and is being used within the callback. 

For example: 

```php
$user = $users->where('id', $this->request->getPost('user_id'))->first();

$activeUsers = $users
    ->where('active', 1)
    ->whenNot($user, function($query, $user) {
        return $query->where('id', '!=', $user->id);
    })
    ->findAll();
```

Ok, this example stinks, also, but I think you see the issue. When passed into `when()`, `$user` becomes a boolean, which cannot be used within the callback.

Reply:
Yes, it is not so simple!
`@param array|bool|float|int|object|resource|string|null $condition`

Reply:
Ah, that makes sense! @iRedds' suggestion makes more sense now where you invert the callables instead of the condition:
```suggestion
        return $this->when($condition, $defaultCallback, $callback);
```

Reply:
That doesn't work either if you haven't specified a default callback, which I predict would be fairly often. 

I think we're stuck with 3 extra lines this time around.

Reply:
I support the current code. It is no problem at all.
It works, and everybody can read and understand.

Adding *null callback* is a bit tricky.

Reply:
👍